Blog: why histograms?

dyladan opened the pull request:
Guest post from https://dyladan.me/histograms/2023/05/02/why-histograms/

This is the first in a series of 3 posts. The second explains the differences and tradeoffs between histograms and summaries (https://dyladan.me/histograms/2023/05/03/histograms-vs-summaries/). The third explains exponential histograms in detail (https://dyladan.me/histograms/2023/05/04/exponential-histograms/).

Please feel free to share any editing or copy suggestions.

---

**Preview**: https://deploy-preview-2658--opentelemetry.netlify.app/blog/2023/why-histograms/

cartermp commented:
FYI - formatting check failed, but since we're probably going to do copy-edits, let's minimize your pain and only have you reformat stuff until after we're satisfied with the content 🙂

dyladan commented:
> FYI - formatting check failed, but since we're probably going to do copy-edits, let's minimize your pain and only have you reformat stuff until after we're satisfied with the content 🙂

It's just as easy to do it now. Reviews may be easier if things are in the expected format.

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+<i>Originally posted to
+https://dyladan.me/histograms/2023/05/02/why-histograms/</i>
+
+A histogram is a multi-value counter which summarizes the distribution of one or
+more data points. For example, a histogram may have 3 counters which count the
+occurrences of negative, positive, and zero values respectively. Given a series
+of numbers, `3`, `-9`, `7`, `6`, `0`, and `-1`, the histogram would count `2`
+negative, `1` zero, and `3` positive values. A single histogram data point is
+most commonly represented as a bar chart.
```
```suggestion
A histogram is a multi-value counter that summarizes the distribution of data points.
For example, a histogram may have 3 counters which count the
occurrences of negative, positive, and zero values respectively. Given a series
of numbers, `3`, `-9`, `7`, `6`, `0`, and `-1`, the histogram would count `2`
negative, `1` zero, and `3` positive values. A single histogram data point is
most commonly represented as a bar chart.
```
slight grammar change and changed "one or more" since people don't use histograms for 1 datapoint, even if it supports it

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+
+![histogram point as bar chart](hist-point.png "A single histogram point plotted as a bar chart with 3 buckets titled 'Positivity of numbers'. The first bucket shows negative numbers and has a height of 2. The second bucket shows zero values and has a height of 1. The third bucket shows positive values and has a height of 3.")
+
+This simple example has only 3 possible output values, but it is not uncommon to
+have many more in a single histogram. Similarly, this shows only 1 histogram
+data point, but in the world of observability data is usually constantly
+exported. For example, an application may export a histogram every minute which
+summarizes a metric for the previous minute. In this way, you can study how the
+distribution of your data changes over time.
```
```suggestion
The above example has only 3 possible output values, but it is common to
have many more in a single histogram. A real-world application typically exports
a histogram every minute that summarizes a metric for the previous minute.
By using histograms this way, you can study how the distribution of your data changes over time.
```

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+
+# Why is that useful?
+
+The most common use-case for histograms in the observability space is defining
+service level objectives (SLOs). One example of such an SLO might be ">=99% of
+all queries should respond in less than 30ms," or "90% of all page loads should
+become interactive within 100ms of first paint."
```
```suggestion
A common use-case for histograms in observability is defining
service level objectives (SLOs). One example of such an SLO might be ">=99% of
all queries should respond in less than 30ms," or "90% of all page loads should
become interactive within 100ms of first paint."
```

cartermp replied:
I don't know if this should be caveated here or not (we'd want to do that in docs though). This point is true when you're working on a metrics data store, but it won't necessarily be true for other tools that let you establish an SLO based on raw data from traces or logs. I'll leave that up to you if you wanna add it or not; I don't think it's urgent to caveat.

dyladan replied:
>This point is true when you're working on a metrics data store, but it won't necessarily be true for other tools that let you establish an SLO based on raw data from traces or logs.

Did you mean to put this on a different line? I'm not sure what claim here might be untrue. It is a common use case for histograms. In any case I like the specific copy edit suggestion you made. "The most" is a bit strong

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+median. More generally, the φ-quantile is the observation value that ranks at
+number φ\*N among the N observations.
+
+# Why is that useful?
```
```suggestion
## Why are Histograms useful?
```

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+summarizes a metric for the previous minute. In this way, you can study how the
+distribution of your data changes over time.
+
+# What are histograms for?
```
```suggestion
## What are histograms for?
```

dyladan replied:
Every heading is H2? Is that a style thing for the Otel blog in general?

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+
+![p99, p90, and p50 plotted as lines](hist-lines.png "p99, p90, and p50 plotted as a line chart with title 'response times.' Time is on the x-axis and response times in milliseconds on the y-axis. p99 response times are around 80 milliseconds. p90 response times are betweeen 60 and 80 milliseconds. p50 response times are between 20 and 30 milliseconds.")
+
+# Other metric types
```
```suggestion
## Other metric types
```

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+collect additional gauges for each percentile, but then you've just force users
+to reimplement histograms on their own. Probably poorly.
+
+# Other data sources and metric types
```
```suggestion
## Other data sources and metric types
```

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+
+# Other metric types
+
+Another solution might be to define the SLOs you're interested in and collect
+them as non-histogram metrics in advance as gauges or counters. This approach
+works, but requires defining your SLOs before you have an understanding of your
+data distribution and requires non-trivial implementation at collection time. It
+is also inflexible; if you decide to change your SLO from 90% of requests to 99%
+of requests, you have to make and release code changes, then wait for the old
+data to age out and the new metric to collect enough data to make useful
+queries. Because histograms model data as a distribution from start to finish,
+they enable you to simply change your queries and get answers on the data you've
+already collected. Particularly with exponential histograms, arbitrary
+distribution queries can be made with very low relative error rates and minimal
+resource consumption on both the client and the analysis backend.
```
```suggestion
What if you're already defining SLOs based on other metrics?
You may have considered defining the SLOs to be based on gauges or counters.
This approach can work, but it requires defining your SLOs before understand your
data distribution and requires non-trivial implementation at collection time. It
is also inflexible; if you decide to change your SLO from 90% of requests to 99%
of requests, you have to make and release code changes, then wait for the old
data to age out and the new metric to collect enough data to make useful
queries. Because histograms model data as a distribution from start to finish,
they enable you to simply change your queries and get answers on the data you've
already collected. Particularly with exponential histograms, arbitrary
distribution queries can be made with very low relative error rates and minimal
resource consumption on both the client and the analysis backend.
```

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+distribution queries can be made with very low relative error rates and minimal
+resource consumption on both the client and the analysis backend.
+
+The inflexibility of this approach also impacts your ability to gauge impact
+when your SLO is violated. For example, imagine you are collecting a gauge which
+calculates the `p99` of some metric and you define an SLO based on it. When your
+SLO is violated and an alert is triggered, how do you know it is really only
+affecting 1% of queries, 10%, or 50%? A histogram would allow you to answer that
+question by simply querying the percentiles you're interested in. You could
+collect additional gauges for each percentile, but then you've just force users
+to reimplement histograms on their own. Probably poorly.
```
```suggestion
The inflexibility of not using histograms for SLOs also impacts your ability to gauge impact
when your SLO is violated. For example, imagine you are collecting a gauge that
calculates the `p99` of some metric and you define an SLO based on it. When your
SLO is violated and an alert is triggered, how do you know it is really only
affecting 1% of queries, 10%, or 50%? A histogram allows you to answer that
question by querying the percentiles you're interested in. You could technically
collect additional gauges for each percentile, but that's just an ad-hoc reimplementation
of histograms anyways, so you're better off using histograms.
```

cartermp commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,86 @@
+
+# Other data sources and metric types
+
+You may ask why you would report a separate metric rather than calculating these
+metrics from your existing log and trace data? While it is true that for _some_
+use cases, like response times, this may be possible, it is not possible for
+_all_ use cases. Even when percentiles can be calculated from existing data,
+this typically requires querying a massive amount of data to answer a very
+simple question. Further, if you are sampling your tracing and logs data, you
+may be missing crucial information required to asses whether or not you are
+meeting your SLOs.
```
```suggestion
You may ask, "why would you report a separate metric rather than calculating these
metrics from your existing log and trace data? While it is true that for _some_
use cases, like response times, this may be possible, it is not necessarily possible for
_all_ use cases. Even when percentiles can be calculated from existing data,
this typically requires querying a massive amount of data to answer a very
simple question. Some observability backends can support this, but others cannot.
Further, if you are sampling your tracing and logs data, you miss crucial information
required to asses whether or not you are meeting your SLOs. Again, this concern is
also backend- and sampling-specific. Depending on your sampling strategy, it may
not be possible to have representative data to accurately calculate an SLO. Further,
if your backend doesn't support re-weighting counts on sampled data, it can
similarly lead to inaccurate SLOs. Using histograms is a way to avoid these subtle
problems if they apply to you.
```

dyladan replied:
What about this:

>You may ask, "why would I report a separate metric rather than calculating it
from my existing log and trace data?" While it is true that for _some_ use
cases, like response times, this may be possible, it is not necessarily possible
for _all_ use cases. Even when quantiles can be calculated from existing data,
you may run into other problems. You need to be sure your observability backend
is able to query and analyze a large amount of existing data on-line or index
and analyze it at ingestion time. If you are sampling your logs and traces or
employing a data retention policy that ages data out, you need to be sure those
things are not affecting derived metrics, or that they are properly re-weighted,
or you risk not being able to accurately asses your SLOs. Depending on your
sampling strategy, it may not even be possible. Using histograms is a way to
avoid these subtle problems if they apply to you.

dyladan replied:
I went out of my way when writing the whole series to try not to mention anything backend specific so I'd like to focus the wording on the underlying principles instead.

cartermp replied:
I love that @dyladan, thanks! Let's go with that and then the post LGTM.

cartermp reviewed (commented):
Looks great overall, thank you! Mostly some small copy-edits, but I modified the last paragraph a a bit to caveat about backends and sampling strategies. Since accurate calculations of SLOs is a super deep topic we probably shouldn't elaborate too much, but I do think we should call out that the problems you can run into won't be the same from backend to backend.

chalin commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,84 @@
+title: Why Histograms?
+date: 2023-05-05
+author: '[Daniel Dyla](https://github.com/dyladan)'
+---
+
+<i>Originally posted to
+https://dyladan.me/histograms/2023/05/02/why-histograms/</i>
```
```suggestion
canonical_url: https://dyladan.me/histograms/2023/05/02/why-histograms/
---

```

chalin replied:
To be consistent with our other posts, please add the following at the end of the page:

```markdown
_A version of this article was [originally posted][] to the author's blog._

[originally posted]: {{% param canonical_url %}}
```

chalin reviewed (commented):
See inline comments, thanks.

chalin commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,84 @@
+response times, but almost all of your users are experiencing response times
+under 90ms.
+
+![p99, p90, and p50 plotted as lines](hist-lines.png "p99, p90, and p50 plotted as a line chart with title 'response times.' Time is on the x-axis and response times in milliseconds on the y-axis. p99 response times are around 80 milliseconds. p90 response times are betweeen 60 and 80 milliseconds. p50 response times are between 20 and 30 milliseconds.")
```
```suggestion
![p99, p90, and p50 plotted as lines](hist-lines.png "p99, p90, and p50 plotted as a line chart with title 'response times.' Time is on the x-axis and response times in milliseconds on the y-axis. p99 response times are around 80 milliseconds. p90 response times are between 60 and 80 milliseconds. p50 response times are between 20 and 30 milliseconds.")
```

chalin commented on content/en/blog/2023/why-histograms/index.md:
```diff
@@ -0,0 +1,84 @@
+---
+title: Why Histograms?
+date: 2023-05-05
+author: '[Daniel Dyla](https://github.com/dyladan)'
```
```suggestion
author: '[Daniel Dyla](https://github.com/dyladan)'
spelling: cSpell:ignore Dyla quantile quantiles reimplementation
```

dyladan commented:
updated the date to monday may 8th to reflect when this is likely to be merged

dyladan commented:
If a maintainer merges this before that obviously they can change the date. I have enabled "allow edits from maintainers" on the PR

dyladan commented:
@cartermp @chalin I rewrote the section "Other metric types" because on re-reading I noticed that the wording was technically incorrect. The last paragraph previously stated that collecting each percentile as a gauge is an ad-hoc implementation of histograms, when this is actually much closer to a summary metric type. I also added a caveat that summaries cannot be aggregated across hosts.

cartermp reviewed (approved):
Thanks! Great content. Unless there's no other objections, I'll merge this on Monday and it'll be live.

chalin reviewed (approved):
👍
